<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_22) on Wed Nov 10 16:53:29 GMT 2010 -->
<TITLE>
GenePubMedScorer
</TITLE>

<META NAME="date" CONTENT="2010-11-10">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="GenePubMedScorer";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/GenePubMedScorer.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;NEXT CLASS</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?gnat/client/nei/GenePubMedScorer.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="GenePubMedScorer.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
gnat.client.nei</FONT>
<BR>
Class GenePubMedScorer</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../resources/inherit.gif" ALT="extended by "><B>gnat.client.nei.GenePubMedScorer</B>
</PRE>
<HR>
<DL>
<DT><PRE>public class <B>GenePubMedScorer</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
Scores one or more genes given their occurrence in a particular piece of text. Its main use
 is to rank a list of gene candidates, depending on which of them is more likely to be referred
 to in the text. 
 <br><br>
 Scoring takes into consideration knowledge about each gene, such as
 <ul>
 <li>Gene Ontology terms associated with it,
 <li>sequence features such as length, domains, mutations,
 <li>protein mass
 <li>origin: species and chromosomal location,
 <li>protein interaction partners,
 <li>diseases associated with a gene,
 <li>etc.
 </ul>
 These information are taken from GO/GOA, EntrezGene, UniProt, and other resources. For each gene,
 each type of information is sought in the given text and compared to the known data, resulting in
 a score and ranking per gene, depending on which gene's information most overlaps with the given
 text.
<P>

<P>
<DL>
<DT><B>Author:</B></DT>
  <DD>Joerg</DD>
</DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#DEBUG_OUTPUT">DEBUG_OUTPUT</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Output format for genes and their scores; 1: plain text, 2: SVM.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;<A HREF="../../../gnat/comparison/GOTermSimilarity.html" title="class in gnat.comparison">GOTermSimilarity</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#goTermScorer">goTermScorer</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#NUMBER_OF_SCORES">NUMBER_OF_SCORES</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The number of sub-scores we currently calculate:<br>
 GO-term, GeneRIF, tissue spec, disease, EntrezGene summary,
 molecular weight, amino acids, chr.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>(package private) &nbsp;java.util.HashMap&lt;java.lang.String,java.util.LinkedList&lt;java.lang.Float&gt;&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#scoreCache">scoreCache</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#verbosity">verbosity</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#GenePubMedScorer(gnat.database.go.GOAccess, java.lang.String)">GenePubMedScorer</A></B>(<A HREF="../../../gnat/database/go/GOAccess.html" title="class in gnat.database.go">GOAccess</A>&nbsp;goAccess,
                 java.lang.String&nbsp;go2gofile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a new GenePubMedScorer with an access to the GO
 term distance database and a file that contains previously
 calculated GO term distances.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#checkContextVector(gnat.representation.ContextVector)">checkContextVector</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Convenience method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#checkContextVectors(gnat.representation.ContextVector, gnat.representation.ContextVector)">checkContextVectors</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                    <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Convenience method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../gnat/representation/Gene.html" title="class in gnat.representation">Gene</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#disambiguate(java.util.Collection, gnat.representation.Text)">disambiguate</A></B>(java.util.Collection&lt;<A HREF="../../../gnat/representation/Gene.html" title="class in gnat.representation">Gene</A>&gt;&nbsp;genes,
             <A HREF="../../../gnat/representation/Text.html" title="class in gnat.representation">Text</A>&nbsp;text)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Decides on a single Gene that best fits the given text.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#finalize()">finalize</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call when not using the Scorer anymore.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getCosineSimilarity(gnat.representation.ContextVector, gnat.representation.ContextVector)">getCosineSimilarity</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                    <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the cosine similarity between both vectors.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getNormalizedOverlapSimilarity(java.util.Collection, gnat.representation.ContextVector)">getNormalizedOverlapSimilarity</A></B>(java.util.Collection&lt;<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&gt;&nbsp;cv1s,
                               <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getNormalizedOverlapSimilarity(gnat.representation.ContextVector, gnat.representation.ContextVector)">getNormalizedOverlapSimilarity</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                               <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_Diseases(java.util.Collection, gnat.representation.ContextVector)">getScore_Diseases</A></B>(java.util.Collection&lt;<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&gt;&nbsp;cv1s,
                  <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_EInteractors(gnat.representation.ContextVector, java.lang.String)">getScore_EInteractors</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                      java.lang.String&nbsp;sourceText)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_Functions(java.util.Collection, gnat.representation.ContextVector)">getScore_Functions</A></B>(java.util.Collection&lt;<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&gt;&nbsp;cv1s,
                   <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_GeneRifs(java.util.Collection, gnat.representation.ContextVector)">getScore_GeneRifs</A></B>(java.util.Collection&lt;<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&gt;&nbsp;cv1s,
                  <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_GOCodes(gnat.representation.ContextVector, gnat.representation.ContextVector)">getScore_GOCodes</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                 <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_Interactors(gnat.representation.ContextVector, gnat.representation.ContextVector, java.lang.String)">getScore_Interactors</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                     <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2,
                     java.lang.String&nbsp;sourceText)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_Keywords(gnat.representation.ContextVector, java.lang.String)">getScore_Keywords</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                  java.lang.String&nbsp;sourceText)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Computes the similarity by comparing the number of keywords found in the given text with
 all keywords assigned to the gene (#found / #all).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_Location(gnat.representation.ContextVector, java.lang.String)">getScore_Location</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                  java.lang.String&nbsp;sourceText)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_ProteinDomain(gnat.representation.ContextVector, java.lang.String)">getScore_ProteinDomain</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                       java.lang.String&nbsp;sourceText)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_ProteinLength(gnat.representation.ContextVector, java.lang.String)">getScore_ProteinLength</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                       java.lang.String&nbsp;sourceText)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_ProteinMass(gnat.representation.ContextVector, java.lang.String)">getScore_ProteinMass</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                     java.lang.String&nbsp;sourceText)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_ProteinMutation(gnat.representation.ContextVector, java.lang.String)">getScore_ProteinMutation</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                         java.lang.String&nbsp;sourceText)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_PubMedID(gnat.representation.ContextVector, int)">getScore_PubMedID</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                  int&nbsp;pmid)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_Summary(gnat.representation.ContextVector, gnat.representation.ContextVector)">getScore_Summary</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                 <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;TODO
 - if the normalized overlap between a text and the gene summary is >0.2, it's almost certainly a hit<br>
 - a normalized overlap of >0.1 should also result in a higher score<br>
 =&gt; stretch normed overlaps [0.0...0.2) -&gt; [0.0...0.9) or something like that<br>
 <br>
 OR: take cosine similarity<br></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_Tissues(java.util.Collection, gnat.representation.ContextVector)">getScore_Tissues</A></B>(java.util.Collection&lt;<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&gt;&nbsp;cv1s,
                 <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore_UInteractors(gnat.representation.ContextVector, java.lang.String)">getScore_UInteractors</A></B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                      java.lang.String&nbsp;sourceText)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScore(gnat.representation.Gene, gnat.representation.Text, java.lang.String)">getScore</A></B>(<A HREF="../../../gnat/representation/Gene.html" title="class in gnat.representation">Gene</A>&nbsp;gene,
         <A HREF="../../../gnat/representation/Text.html" title="class in gnat.representation">Text</A>&nbsp;text,
         java.lang.String&nbsp;geneName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculates a score for the Gene in the context of the Text.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.LinkedList&lt;java.lang.Float&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#getScores(gnat.representation.Gene, gnat.representation.Text, java.lang.String)">getScores</A></B>(<A HREF="../../../gnat/representation/Gene.html" title="class in gnat.representation">Gene</A>&nbsp;gene,
          <A HREF="../../../gnat/representation/Text.html" title="class in gnat.representation">Text</A>&nbsp;text,
          java.lang.String&nbsp;geneName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a list of scores for scoring each context vector.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#main(java.lang.String[])">main</A></B>(java.lang.String[]&nbsp;args)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;For testing purposes</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#setVerbosity(int)">setVerbosity</A></B>(int&nbsp;verbosity)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../gnat/client/nei/GenePubMedScorer.html#test()">test</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="goTermScorer"><!-- --></A><H3>
goTermScorer</H3>
<PRE>
private <A HREF="../../../gnat/comparison/GOTermSimilarity.html" title="class in gnat.comparison">GOTermSimilarity</A> <B>goTermScorer</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="verbosity"><!-- --></A><H3>
verbosity</H3>
<PRE>
public int <B>verbosity</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="NUMBER_OF_SCORES"><!-- --></A><H3>
NUMBER_OF_SCORES</H3>
<PRE>
public final int <B>NUMBER_OF_SCORES</B></PRE>
<DL>
<DD>The number of sub-scores we currently calculate:<br>
 GO-term, GeneRIF, tissue spec, disease, EntrezGene summary,
 molecular weight, amino acids, chr. location, mutations,
 function, keywords, PubMed, protein domains.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../constant-values.html#gnat.client.nei.GenePubMedScorer.NUMBER_OF_SCORES">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="DEBUG_OUTPUT"><!-- --></A><H3>
DEBUG_OUTPUT</H3>
<PRE>
public int <B>DEBUG_OUTPUT</B></PRE>
<DL>
<DD>Output format for genes and their scores; 1: plain text, 2: SVM.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="scoreCache"><!-- --></A><H3>
scoreCache</H3>
<PRE>
java.util.HashMap&lt;java.lang.String,java.util.LinkedList&lt;java.lang.Float&gt;&gt; <B>scoreCache</B></PRE>
<DL>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="GenePubMedScorer(gnat.database.go.GOAccess, java.lang.String)"><!-- --></A><H3>
GenePubMedScorer</H3>
<PRE>
public <B>GenePubMedScorer</B>(<A HREF="../../../gnat/database/go/GOAccess.html" title="class in gnat.database.go">GOAccess</A>&nbsp;goAccess,
                        java.lang.String&nbsp;go2gofile)</PRE>
<DL>
<DD>Constructs a new GenePubMedScorer with an access to the GO
 term distance database and a file that contains previously
 calculated GO term distances.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>goAccess</CODE> - <DD><CODE>go2gofile</CODE> - </DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="finalize()"><!-- --></A><H3>
finalize</H3>
<PRE>
public void <B>finalize</B>()</PRE>
<DL>
<DD>Call when not using the Scorer anymore. Closes the DB connection
 and writes new GO distances to the file <tt>go2gofile</tt>.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>finalize</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="checkContextVector(gnat.representation.ContextVector)"><!-- --></A><H3>
checkContextVector</H3>
<PRE>
public boolean <B>checkContextVector</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv)</PRE>
<DL>
<DD>Convenience method. Returns true if cv is not null.
<P>
<DD><DL>

<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="checkContextVectors(gnat.representation.ContextVector, gnat.representation.ContextVector)"><!-- --></A><H3>
checkContextVectors</H3>
<PRE>
public boolean <B>checkContextVectors</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                                   <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</PRE>
<DL>
<DD>Convenience method. Returns true if both vectors check to true.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>cv2</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="disambiguate(java.util.Collection, gnat.representation.Text)"><!-- --></A><H3>
disambiguate</H3>
<PRE>
public <A HREF="../../../gnat/representation/Gene.html" title="class in gnat.representation">Gene</A> <B>disambiguate</B>(java.util.Collection&lt;<A HREF="../../../gnat/representation/Gene.html" title="class in gnat.representation">Gene</A>&gt;&nbsp;genes,
                         <A HREF="../../../gnat/representation/Text.html" title="class in gnat.representation">Text</A>&nbsp;text)</PRE>
<DL>
<DD>Decides on a single Gene that best fits the given text. Decision is
 based on score (average and number of scored elements [GO terms,
 diseases, GeneRIFS, etc.]).
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>genes</CODE> - <DD><CODE>text</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getCosineSimilarity(gnat.representation.ContextVector, gnat.representation.ContextVector)"><!-- --></A><H3>
getCosineSimilarity</H3>
<PRE>
public float <B>getCosineSimilarity</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                                 <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</PRE>
<DL>
<DD>Returns the cosine similarity between both vectors.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>cv2</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getNormalizedOverlapSimilarity(gnat.representation.ContextVector, gnat.representation.ContextVector)"><!-- --></A><H3>
getNormalizedOverlapSimilarity</H3>
<PRE>
public float <B>getNormalizedOverlapSimilarity</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                                            <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>cv2</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getNormalizedOverlapSimilarity(java.util.Collection, gnat.representation.ContextVector)"><!-- --></A><H3>
getNormalizedOverlapSimilarity</H3>
<PRE>
public float <B>getNormalizedOverlapSimilarity</B>(java.util.Collection&lt;<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&gt;&nbsp;cv1s,
                                            <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1s</CODE> - <DD><CODE>cv2</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore(gnat.representation.Gene, gnat.representation.Text, java.lang.String)"><!-- --></A><H3>
getScore</H3>
<PRE>
public float <B>getScore</B>(<A HREF="../../../gnat/representation/Gene.html" title="class in gnat.representation">Gene</A>&nbsp;gene,
                      <A HREF="../../../gnat/representation/Text.html" title="class in gnat.representation">Text</A>&nbsp;text,
                      java.lang.String&nbsp;geneName)</PRE>
<DL>
<DD>Calculates a score for the Gene in the context of the Text.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>gene</CODE> - <DD><CODE>text</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScores(gnat.representation.Gene, gnat.representation.Text, java.lang.String)"><!-- --></A><H3>
getScores</H3>
<PRE>
public java.util.LinkedList&lt;java.lang.Float&gt; <B>getScores</B>(<A HREF="../../../gnat/representation/Gene.html" title="class in gnat.representation">Gene</A>&nbsp;gene,
                                                       <A HREF="../../../gnat/representation/Text.html" title="class in gnat.representation">Text</A>&nbsp;text,
                                                       java.lang.String&nbsp;geneName)</PRE>
<DL>
<DD>Returns a list of scores for scoring each context vector. Each score is between -1.0f and 1.0f.
 Some entries in the list might be null, when the gene did not have a corresponding context vector;
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>gene</CODE> - <DD><CODE>text</CODE> - <DD><CODE>geneName</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_Diseases(java.util.Collection, gnat.representation.ContextVector)"><!-- --></A><H3>
getScore_Diseases</H3>
<PRE>
public float <B>getScore_Diseases</B>(java.util.Collection&lt;<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&gt;&nbsp;cv1s,
                               <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1s</CODE> - <DD><CODE>cv2</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_Functions(java.util.Collection, gnat.representation.ContextVector)"><!-- --></A><H3>
getScore_Functions</H3>
<PRE>
public float <B>getScore_Functions</B>(java.util.Collection&lt;<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&gt;&nbsp;cv1s,
                                <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1s</CODE> - <DD><CODE>cv2</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_GeneRifs(java.util.Collection, gnat.representation.ContextVector)"><!-- --></A><H3>
getScore_GeneRifs</H3>
<PRE>
public float <B>getScore_GeneRifs</B>(java.util.Collection&lt;<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&gt;&nbsp;cv1s,
                               <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1s</CODE> - <DD><CODE>cv2</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_GOCodes(gnat.representation.ContextVector, gnat.representation.ContextVector)"><!-- --></A><H3>
getScore_GOCodes</H3>
<PRE>
public float <B>getScore_GOCodes</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                              <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_Keywords(gnat.representation.ContextVector, java.lang.String)"><!-- --></A><H3>
getScore_Keywords</H3>
<PRE>
public float <B>getScore_Keywords</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                               java.lang.String&nbsp;sourceText)</PRE>
<DL>
<DD>Computes the similarity by comparing the number of keywords found in the given text with
 all keywords assigned to the gene (#found / #all).
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>sourceText</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_Location(gnat.representation.ContextVector, java.lang.String)"><!-- --></A><H3>
getScore_Location</H3>
<PRE>
public float <B>getScore_Location</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                               java.lang.String&nbsp;sourceText)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>sourceText</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_ProteinDomain(gnat.representation.ContextVector, java.lang.String)"><!-- --></A><H3>
getScore_ProteinDomain</H3>
<PRE>
public float <B>getScore_ProteinDomain</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                                    java.lang.String&nbsp;sourceText)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>sourceText</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_ProteinLength(gnat.representation.ContextVector, java.lang.String)"><!-- --></A><H3>
getScore_ProteinLength</H3>
<PRE>
public float <B>getScore_ProteinLength</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                                    java.lang.String&nbsp;sourceText)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>sourceText</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_ProteinMass(gnat.representation.ContextVector, java.lang.String)"><!-- --></A><H3>
getScore_ProteinMass</H3>
<PRE>
public float <B>getScore_ProteinMass</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                                  java.lang.String&nbsp;sourceText)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>sourceText</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_ProteinMutation(gnat.representation.ContextVector, java.lang.String)"><!-- --></A><H3>
getScore_ProteinMutation</H3>
<PRE>
public float <B>getScore_ProteinMutation</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                                      java.lang.String&nbsp;sourceText)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>sourceText</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_EInteractors(gnat.representation.ContextVector, java.lang.String)"><!-- --></A><H3>
getScore_EInteractors</H3>
<PRE>
public float <B>getScore_EInteractors</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                                   java.lang.String&nbsp;sourceText)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>sourceText</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_UInteractors(gnat.representation.ContextVector, java.lang.String)"><!-- --></A><H3>
getScore_UInteractors</H3>
<PRE>
public float <B>getScore_UInteractors</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                                   java.lang.String&nbsp;sourceText)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>sourceText</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_Interactors(gnat.representation.ContextVector, gnat.representation.ContextVector, java.lang.String)"><!-- --></A><H3>
getScore_Interactors</H3>
<PRE>
public float <B>getScore_Interactors</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                                  <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2,
                                  java.lang.String&nbsp;sourceText)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>cv2</CODE> - <DD><CODE>sourceText</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_PubMedID(gnat.representation.ContextVector, int)"><!-- --></A><H3>
getScore_PubMedID</H3>
<PRE>
public float <B>getScore_PubMedID</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                               int&nbsp;pmid)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>pmid</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_Summary(gnat.representation.ContextVector, gnat.representation.ContextVector)"><!-- --></A><H3>
getScore_Summary</H3>
<PRE>
public float <B>getScore_Summary</B>(<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv1,
                              <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</PRE>
<DL>
<DD>TODO
 - if the normalized overlap between a text and the gene summary is >0.2, it's almost certainly a hit<br>
 - a normalized overlap of >0.1 should also result in a higher score<br>
 =&gt; stretch normed overlaps [0.0...0.2) -&gt; [0.0...0.9) or something like that<br>
 <br>
 OR: take cosine similarity<br>
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1</CODE> - <DD><CODE>cv2</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="getScore_Tissues(java.util.Collection, gnat.representation.ContextVector)"><!-- --></A><H3>
getScore_Tissues</H3>
<PRE>
public float <B>getScore_Tissues</B>(java.util.Collection&lt;<A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&gt;&nbsp;cv1s,
                              <A HREF="../../../gnat/representation/ContextVector.html" title="class in gnat.representation">ContextVector</A>&nbsp;cv2)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cv1s</CODE> - <DD><CODE>cv2</CODE> - 
<DT><B>Returns:</B><DD></DL>
</DD>
</DL>
<HR>

<A NAME="test()"><!-- --></A><H3>
test</H3>
<PRE>
public static void <B>test</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setVerbosity(int)"><!-- --></A><H3>
setVerbosity</H3>
<PRE>
public void <B>setVerbosity</B>(int&nbsp;verbosity)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>verbosity</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="main(java.lang.String[])"><!-- --></A><H3>
main</H3>
<PRE>
public static void <B>main</B>(java.lang.String[]&nbsp;args)
                 throws java.lang.Exception</PRE>
<DL>
<DD>For testing purposes
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>args</CODE> - 
<DT><B>Throws:</B>
<DD><CODE>java.lang.Exception</CODE></DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/GenePubMedScorer.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;NEXT CLASS</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?gnat/client/nei/GenePubMedScorer.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="GenePubMedScorer.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
